File: Server/Src/Jamming/Handler/JammerHandler.cs
using System.Text.Json;

public class JammerHandler
{
    private static JammerHandler instance;
    private readonly JammerManager jammerManager = JammerManager.GetInstance();
    private readonly JammersDataManager jammersDataManager = JammersDataManager.GetInstance();

    private JammerHandler()
    {
    }

    public static JammerHandler GetInstance()
    {
        if (instance == null)
            instance = new JammerHandler();
        return instance;
    }

    public void HandleAddJammer(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Jammer jammer = JsonSerializer.Deserialize<Jammer>(data);

            // unique ID
            Guid uuid = Guid.NewGuid();
            jammer.id = uuid.ToString();

            // add to file/db
            jammersDataManager.AddAndSaveJammer(jammer);

            // add to manager map
            bool isAdded = jammerManager.TryAddJammer(jammer);
            if (isAdded)
            {
                System.Console.WriteLine("{0} ({1}) - Added jammer successfully.", jammer.id, jammer.id);
                SendAddJammer(jammer, clientMode);
            }
            else
            {
                System.Console.WriteLine("{0} - Failed to add jammer.", jammer.id);
                SendJammerError($"{jammer.id} - Failed to add jammer.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleAddJammer: " + ex.Message);
        }
    }

    public void HandleRemoveJammer(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Jammer jammer = JsonSerializer.Deserialize<Jammer>(data);
            string jammerId = jammer.id;

            bool isRemoved = jammersDataManager.RemoveAndSaveJammer(jammerId);
            if (isRemoved)
            {
                isRemoved = jammerManager.TryRemoveJammer(jammerId);
                if (isRemoved)
                {
                    System.Console.WriteLine("{0} - Removed jammer successfully.", jammerId);
                    SendRemoveJammer(jammer, clientMode);
                }
                else
                {
                    System.Console.WriteLine("{0} - Failed to remove jammer from manager.", jammerId);
                    SendJammerError($"{jammerId} - Failed to remove jammer from manager.", clientMode);
                }
            }
            else
            {
                System.Console.WriteLine("{0} - Failed to remove jammer from file.", jammerId);
                SendJammerError($"{jammerId} - Failed to remove jammer from file.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleRemoveJammer: " + ex.Message);
        }
    }

    public void HandleEditJammer(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Jammer jammer = JsonSerializer.Deserialize<Jammer>(data);
            string jammerId = jammer.id;

            bool isEdited = jammersDataManager.EditAndSaveJammer(jammerId, jammer);
            if (isEdited)
            {
                isEdited = jammerManager.TryEditJammer(jammerId, jammer);
                if (isEdited)
                {
                    System.Console.WriteLine("{0} - Edited jammer successfully.", jammerId);
                    SendEditJammer(jammer, clientMode);
                }
                else
                {
                    System.Console.WriteLine("{0} - Failed to edit jammer in manager.", jammerId);
                    SendJammerError($"{jammerId} - Failed to edit jammer in manager.", clientMode);
                }
            }
            else
            {
                System.Console.WriteLine("{0} - Failed to edit jammer in file.", jammerId);
                SendJammerError($"{jammerId} - Failed to edit jammer in file.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleEditJammer: " + ex.Message);
        }
    }

    public void SendAddJammer(Jammer jammer, ModeEnum clientMode)
    {
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.AddJammer, jammer, clientMode);
        WebSocketServer.SendMsgToClients(data, clientMode);
    }

    public void SendRemoveJammer(Jammer jammer, ModeEnum clientMode)
    {
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.RemoveJammer, jammer, clientMode);
        WebSocketServer.SendMsgToClients(data, clientMode);
    }

    public void SendEditJammer(Jammer jammer, ModeEnum clientMode)
    {
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.EditJammer, jammer, clientMode);
        WebSocketServer.SendMsgToClients(data, clientMode);
    }

    public void SendJammerError(string errorMsg, ModeEnum clientMode)
    {
        JammerError error = new JammerError() { errorMsg = errorMsg };
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.JammerError, error, clientMode);
        WebSocketServer.SendMsgToClients(data, clientMode);
    }
}

File: C2ReactClient/src/Jamming/Handler/JammerHandler.ts
import * as Cesium from "cesium";
import { toast } from "react-toastify";
import { JammersManager } from "../Manager/JammerManager";
import { JammerEntityManager } from "../EntitiesManagment/JammerEntityManager";
import type { Jammer } from "../Jammer/Jammer";

export class JammerHandler {
  private static instance: JammerHandler | null = null;
  private jammerManager: JammersManager;
  private jammerEntityManager: JammerEntityManager;

  private constructor(viewer?: Cesium.Viewer) {
    this.jammerManager = JammersManager.getInstance();
    this.jammerEntityManager = JammerEntityManager.GetInstance(viewer!);
  }

  public static getInstance(viewer?: Cesium.Viewer): JammerHandler {
    if (this.instance === null) {
      this.instance = new JammerHandler(viewer);
    }
    return this.instance;
  }

  HandleAddJammer(data: any) {
    try {
      const jammer = data as Jammer;
      this.AddJammer(jammer);
    } catch (err) {
      console.log("data could not be parsed to Jammer");
    }
  }

  AddJammer(jammer: Jammer) {
    const isAdded = this.jammerManager.tryAddJammer(jammer);
    if (isAdded) {
      this.jammerEntityManager.tryAddJammer(jammer);
      console.log(`Jammer ${jammer.id} added successfully.`);
    } else {
      console.log("error in HandleAddJammer. Jammer adding failed");
    }
  }

  HandleRemoveJammer(data: any) {
    try {
      const jammer = data as Jammer;
      this.RemoveJammer(jammer);
    } catch (err) {
      console.log("data could not be parsed to Jammer");
    }
  }

  RemoveJammer(jammer: Jammer) {
    const isRemoved = this.jammerManager.tryRemoveJammer(jammer.id);
    if (isRemoved) {
      this.jammerEntityManager.removeJammer(jammer.id);
      console.log(`Jammer ${jammer.id} removed successfully.`);
    } else {
      console.log("error in HandleRemoveJammer. Jammer removing failed");
    }
  }

  HandleEditJammer(data: any) {
    try {
      const jammer = data as Jammer;
      this.EditJammer(jammer);
    } catch (err) {
      console.log("data could not be parsed to Jammer");
    }
  }

  EditJammer(jammer: Jammer) {
    const isEdited = this.jammerManager.tryEditJammer(jammer);
    if (isEdited) {
      this.jammerEntityManager.editJammer(jammer);
      console.log(`Jammer ${jammer.id} edited successfully.`);
    } else {
      console.log("error in HandleEditJammer. Jammer editing failed");
    }
  }
}